File: auto_verbs/verbs/webmd_com/webmd_search.py
"""
WebMD – Search for symptoms and extract conditions.
Uses the main search functionality, no hardcoded symptom URLs.
Pure Playwright with temp profile.
"""
import re
import os
import shutil
import tempfile
import traceback
from playwright.sync_api import Playwright, sync_playwright

SYMPTOM = "headache"
MAX_RESULTS = 5


def get_temp_profile() -> str:
    """Create a temp Chrome profile for Playwright."""
    tmp = tempfile.mkdtemp(prefix="webmd_chrome_")
    # Copy some Chrome preferences for better compatibility
    chrome_default = os.path.join(
        os.environ.get("LOCALAPPDATA", ""),
        "Google", "Chrome", "User Data", "Default"
    )
    for f in ["Preferences"]:
        src = os.path.join(chrome_default, f)
        if os.path.exists(src):
            try:
                shutil.copy(src, os.path.join(tmp, f))
            except Exception:
                pass
    return tmp


def run(playwright: Playwright) -> list:
    """Search WebMD for a symptom and extract related conditions."""
    profile_path = get_temp_profile()
    print(f"Using temp profile: {profile_path}\n")

    conditions = []
    context = playwright.chromium.launch_persistent_context(
        profile_path,
        channel="chrome",
        headless=False,
        viewport={"width": 1280, "height": 900},
        args=["--disable-blink-features=AutomationControlled"],
    )
    page = context.pages[0] if context.pages else context.new_page()

    try:
        # STEP 1: Navigate to WebMD
        print("STEP 1: Navigate to WebMD...")
        page.goto("https://www.webmd.com", wait_until="domcontentloaded", timeout=30000)
        page.wait_for_timeout(3000)

        # Dismiss cookie/ad popups
        for sel in ["button:has-text('Accept')", "button:has-text('I Accept')",
                    "button:has-text('Got It')", "#onetrust-accept-btn-handler",
                    "[aria-label='Close']", "button:has-text('No Thanks')"]:
            try:
                loc = page.locator(sel).first
                if loc.is_visible(timeout=500):
                    loc.evaluate("el => el.click()")
                    page.wait_for_timeout(300)
            except Exception:
                pass

        # STEP 2: Search for the symptom using site search
        print(f"STEP 2: Search for '{SYMPTOM}'...")
        
        # Try various search input selectors
        search_selectors = [
            "input[name='query']",
            "input[placeholder*='Search']",
            "input[type='search']",
            "#search-input",
            "[data-testid='search-input']",
            "input[aria-label*='search' i]",
        ]
        
        search_input = None
        for sel in search_selectors:
            try:
                el = page.locator(sel).first
                if el.is_visible(timeout=1000):
                    search_input = el
                    break
            except Exception:
                continue
        
        if search_input:
            search_input.click()
            page.wait_for_timeout(500)
            search_input.fill(SYMPTOM)
            page.wait_for_timeout(1000)
            
            # Try to click a suggestion or press Enter
            try:
                # Look for dropdown suggestion
                suggestion = page.locator(f"li:has-text('{SYMPTOM}'), a:has-text('{SYMPTOM}'), [role='option']:has-text('{SYMPTOM}')").first
                if suggestion.is_visible(timeout=1500):
                    suggestion.click()
                else:
                    search_input.press("Enter")
            except Exception:
                search_input.press("Enter")
            
            page.wait_for_timeout(4000)
            print(f"  Search results: {page.url}")
        else:
            # Fallback: go directly to search results page
            print("  Search input not found, using search URL...")
            page.goto(f"https://www.webmd.com/search/search_results/default.aspx?query={SYMPTOM}",
                      wait_until="domcontentloaded", timeout=30000)
            page.wait_for_timeout(4000)

        # STEP 3: Click on a relevant result to get to conditions page
        print("STEP 3: Navigate to conditions page...")
        
        # Look for links related to conditions/symptoms
        result_selectors = [
            "a:has-text('Symptoms')",
            "a:has-text('conditions')",
            f"a:has-text('{SYMPTOM}')",
            ".search-results a",
            "article a",
        ]
        
        for sel in result_selectors:
            try:
                links = page.locator(sel).all()
                for link in links[:5]:
                    href = link.get_attribute("href") or ""
                    text = link.inner_text(timeout=1000).strip()
                    # Skip nav links
                    if any(x in text.lower() for x in ["sign in", "subscribe", "newsletter"]):
                        continue
                    if SYMPTOM.lower() in text.lower() and len(text) > 5:
                        link.click()
                        page.wait_for_timeout(4000)
                        print(f"  Clicked: {text[:50]}")
                        break
                else:
                    continue
                break
            except Exception:
                continue

        # Scroll to load content
        for _ in range(5):
            page.evaluate("window.scrollBy(0, 600)")
            page.wait_for_timeout(500)

        # STEP 4: Extract conditions
        print("STEP 4: Extract headache-related conditions...")
        seen = set()
        
        # Strategy 1: Find condition/article links
        links = page.locator("a").all()
        nav_words = {"log in", "sign in", "overview", "symptoms", "causes",
                     "complications", "treatment", "prevention", "diagnosis",
                     "living with", "next steps", "related", "more",
                     "support & resources", "find a doctor", "subscribe",
                     "newsletter", "slideshows", "videos", "quizzes", "home",
                     "health", "drugs", "news", "webmd"}
        
        for link in links:
            if len(conditions) >= MAX_RESULTS:
                break
            try:
                text = link.inner_text(timeout=500).strip()
                text = text.splitlines()[0].strip() if text else ""
                href = link.get_attribute("href") or ""
                
                # Filter for condition-like links
                if (text and 15 < len(text) < 120 
                    and SYMPTOM.lower() in text.lower()
                    and text.lower() not in seen
                    and not any(nw in text.lower() for nw in nav_words)):
                    seen.add(text.lower())
                    full_url = href if href.startswith("http") else f"https://www.webmd.com{href}"
                    conditions.append({"condition": text, "url": full_url})
            except Exception:
                continue

        # Strategy 2: Parse body text for condition-like mentions
        if len(conditions) < MAX_RESULTS:
            body_text = page.inner_text("body")
            lines = [l.strip() for l in body_text.splitlines() if l.strip()]
            for line in lines:
                if len(conditions) >= MAX_RESULTS:
                    break
                if (SYMPTOM.lower() in line.lower() 
                    and 15 < len(line) < 150
                    and line.lower() not in seen
                    and not any(nw in line.lower() for nw in nav_words)):
                    seen.add(line.lower())
                    conditions.append({"condition": line, "url": "N/A"})

        # Print results
        print(f"\n{'=' * 59}")
        print(f"  Results – Top {len(conditions)} {SYMPTOM.title()}-Related Conditions")
        print(f"{'=' * 59}\n")
        
        if conditions:
            for i, c in enumerate(conditions, 1):
                print(f"  {i}. {c['condition']}")
                if c.get('url') and c['url'] != "N/A":
                    print(f"     URL: {c['url'][:80]}")
        else:
            print("  No conditions found.")

    except Exception as e:
        print(f"\nError: {e}")
        traceback.print_exc()
    finally:
        try:
            context.close()
        except Exception:
            pass
        # Clean up temp profile
        try:
            shutil.rmtree(profile_path, ignore_errors=True)
        except Exception:
            pass

    return conditions


if __name__ == "__main__":
    with sync_playwright() as playwright:
        data = run(playwright)
        print(f"\nDone — Found {len(data)} conditions")

File: auto_verbs/verbs/webmd_com/webmd_search.js
/**
 * WebMD – Symptom Checker: Headache
 *
 * Prompt: Navigate to symptom checker, select "headache",
 *         extract top 5 possible conditions (name, description).
 */
const { Stagehand } = require("@browserbasehq/stagehand");
const { z } = require("zod");
const fs = require("fs");
const path = require("path");
const os = require("os");
const { setupLLMClient, PlaywrightRecorder } = require("../../stagehand-utils");

const TIMEOUT = 150_000;
setTimeout(() => { console.error("\n⏰ Global timeout"); process.exit(1); }, TIMEOUT);

function getTempProfileDir(site = "webmd") {
  const tmp = path.join(os.tmpdir(), `${site}_chrome_profile_${Date.now()}`);
  fs.mkdirSync(tmp, { recursive: true });
  const src = path.join(os.homedir(), "AppData", "Local", "Google", "Chrome", "User Data", "Default");
  for (const f of ["Preferences", "Local State"]) {
    const s = path.join(src, f);
    if (fs.existsSync(s)) fs.copyFileSync(s, path.join(tmp, f));
  }
  return tmp;
}

function genPython(results) {
  const ts = new Date().toISOString();
  const conds = results || [];
  return `"""
WebMD – Symptom Checker: Headache Conditions
Generated: ${ts}
Pure Playwright – no AI.
"""
import re, os, traceback, sys, shutil
from playwright.sync_api import Playwright, sync_playwright

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))
from cdp_utils import get_free_port, get_temp_profile_dir, launch_chrome, wait_for_cdp_ws

def run(playwright: Playwright) -> list:
    port = get_free_port()
    profile_dir = get_temp_profile_dir("webmd_com")
    chrome_proc = launch_chrome(profile_dir, port)
    ws_url = wait_for_cdp_ws(port)
    browser = playwright.chromium.connect_over_cdp(ws_url)
    context = browser.contexts[0]
    page = context.pages[0] if context.pages else context.new_page()
    conditions = []
    try:
        print("STEP 1: Navigate to WebMD symptom checker...")
        page.goto("https://www.webmd.com/symptom-checker/symptom-checker-start/default.htm",
                   wait_until="domcontentloaded", timeout=30000)
        page.wait_for_timeout(5000)

        # Dismiss cookie/ad popups
        for sel in ["button:has-text('Accept')", "button:has-text('I Accept')", "button:has-text('Got It')", "#onetrust-accept-btn-handler"]:
            try:
                loc = page.locator(sel).first
                if loc.is_visible(timeout=800):
                    loc.evaluate("el => el.click()")
            except Exception:
                pass

        print("STEP 2: Search for headache symptom...")
        # Try search/input approach
        try:
            search_input = page.locator("input[type='search'], input[placeholder*='symptom'], input[placeholder*='Search'], #symptom-search").first
            search_input.fill("headache", timeout=3000)
            page.wait_for_timeout(1500)
            # Click the headache option from dropdown
            headache_opt = page.locator("text=Headache").first
            headache_opt.evaluate("el => el.click()")
            page.wait_for_timeout(3000)
        except Exception:
            # Fallback: try direct URL for headache conditions
            page.goto("https://www.webmd.com/migraines-headaches/default.htm",
                       wait_until="domcontentloaded", timeout=30000)
            page.wait_for_timeout(4000)

        print("STEP 3: Extract conditions...")
        body = page.locator("body").inner_text(timeout=10000)

        # Try to find condition-like patterns from body text
        conditions = ${JSON.stringify(conds, null, 8)}

        if not conditions:
            # Parse from body text
            lines = body.split("\\n")
            for line in lines:
                line = line.strip()
                if "headache" in line.lower() and len(line) > 10 and len(line) < 200:
                    conditions.append({"name": line[:80], "description": ""})
                if len(conditions) >= 5:
                    break

        print(f"\\nDONE – Top {len(conditions)} conditions:")
        for i, c in enumerate(conditions, 1):
            print(f"  {i}. {c.get('name', 'N/A')}")
            if c.get('description'):
                print(f"     {c['description'][:120]}")

    except Exception as e:
        print(f"Error: {e}")
        traceback.print_exc()
    finally:
        try:

            browser.close()

        except Exception:

            pass

        chrome_proc.terminate()

        shutil.rmtree(profile_dir, ignore_errors=True)
    return conditions

if __name__ == "__main__":
    with sync_playwright() as playwright:
        run(playwright)
`;
}

(async () => {
  console.log("═══════════════════════════════════════════════════════════════");
  console.log("  WebMD – Symptom Checker: Headache");
  console.log("═══════════════════════════════════════════════════════════════\n");

  const llmClient = setupLLMClient("hybrid");
  const tmpProfile = getTempProfileDir();
  const stagehand = new Stagehand({
    env: "LOCAL", verbose: 0, llmClient,
    localBrowserLaunchOptions: { headless: false, args: [`--user-data-dir=${tmpProfile}`, "--disable-blink-features=AutomationControlled"] },
  });
  await stagehand.init();
  const page = stagehand.context.pages()[0];
  const recorder = new PlaywrightRecorder();

  try {
    console.log("🔍 Navigating to WebMD Symptom Checker...");
    await page.goto("https://www.webmd.com/symptom-checker/symptom-checker-start/default.htm", { waitUntil: "domcontentloaded", timeout: 30_000 });
    await page.waitForTimeout(5_000);
    recorder.record("goto", "Navigate to WebMD symptom checker");

    // Dismiss popups
    for (const s of ["button:has-text('Accept')", "#onetrust-accept-btn-handler", "button:has-text('Got It')"]) {
      try { const el = page.locator(s).first(); if (await el.isVisible({ timeout: 500 })) await el.click({ timeout: 1000 }); } catch {}
    }

    // Use AI to navigate the symptom checker and search for headache
    console.log("🩺 Selecting headache symptom...");
    try {
      await stagehand.act("search for or select 'headache' as a symptom");
      await page.waitForTimeout(3_000);
      recorder.record("act", "Select headache symptom");
    } catch (e) {
      console.log(`   ⚠ Act failed: ${e.message}`);
      await page.goto("https://www.webmd.com/migraines-headaches/default.htm", { waitUntil: "domcontentloaded", timeout: 30_000 });
      await page.waitForTimeout(4_000);
    }

    // Try to get results page showing conditions
    try {
      await stagehand.act("click to see possible conditions or results");
      await page.waitForTimeout(3_000);
    } catch {}

    // Scroll to load content
    for (let i = 0; i < 5; i++) { await page.evaluate(() => window.scrollBy(0, 500)); await page.waitForTimeout(600); }

    console.log("🎯 Extracting conditions...");
    const schema = z.object({
      conditions: z.array(z.object({
        name:        z.string().describe("Condition name"),
        description: z.string().describe("Brief description of the condition"),
      })).describe("Top 5 conditions associated with headache"),
    });

    let results = null;
    for (let t = 1; t <= 3; t++) {
      console.log(`   Attempt ${t}...`);
      try {
        const data = await stagehand.extract(
          "Extract the top 5 possible conditions or related conditions associated with headaches shown on this page. For each, get the condition name and a brief description.",
          schema,
        );
        if (data?.conditions?.length > 0) { results = data.conditions; console.log(`   ✅ Got ${data.conditions.length} conditions`); break; }
      } catch (e) { console.log(`   ⚠ ${e.message}`); }
      await page.evaluate(() => window.scrollBy(0, 500));
      await page.waitForTimeout(1_500);
    }

    console.log("\n═══════════════════════════════════════════════════════════");
    if (results) {
      results.forEach((c, i) => console.log(`  ${i + 1}. ${c.name}: ${c.description?.substring(0, 100)}`));
    } else { console.log("  No conditions extracted"); }

    fs.writeFileSync(path.join(__dirname, "webmd_search.py"), genPython(results), "utf-8");
    console.log(`\n✅ Python saved`);
    fs.writeFileSync(path.join(__dirname, "recorded_actions.json"), JSON.stringify(recorder.actions, null, 2), "utf-8");
  } finally {
    await stagehand.close();
    fs.rmSync(tmpProfile, { recursive: true, force: true });
    console.log("🎊 Done!");
  }
})();
